File: apps/web/src/lib/api/recovery.ts
/**
 * Cart Recovery API functions
 */

import { apiGet, apiPatch, apiPost } from './client';
import type { PaginatedResponse } from './types';

// === Types ===

export interface RecoverySummary {
  total_abandoned: number;
  total_recovered: number;
  recovery_rate: number;
  recovered_revenue: number;
  emails_sent: number;
  active_sequences: number;
  period_days: number;
}

export interface RecoveryDailyCount {
  date: string;
  abandoned: number;
  recovered: number;
}

export interface RecoverySequence {
  id: string;
  abandoned_checkout_id: string;
  customer_email: string;
  sequence_type: string;
  status: string;
  current_step_index: number;
  steps_completed: Record<string, unknown>[];
  total_steps: number;
  next_step_at: string | null;
  started_at: string;
  completed_at: string | null;
  created_at: string;
}

export interface AbandonedCheckout {
  id: string;
  shopify_checkout_id: string;
  customer_email: string | null;
  customer_name: string | null;
  total_price: number;
  currency: string;
  line_items: Record<string, unknown>[];
  checkout_url: string | null;
  status: string;
  abandonment_detected_at: string | null;
  recovered_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecoverySettings {
  enabled: boolean;
  min_cart_value: number;
  abandonment_threshold_minutes: number;
  sequence_timing_minutes: number[];
  discount_enabled: boolean;
  discount_percent: number;
  max_emails_per_day: number;
  exclude_email_patterns: string[];
}

// === Query Functions ===

export async function getRecoverySummary(
  storeId: string,
  days?: number
): Promise<RecoverySummary> {
  return apiGet<RecoverySummary>('/api/v1/recovery/analytics/summary', {
    params: { store_id: storeId, ...(days ? { days: String(days) } : {}) },
  });
}

export async function getRecoveryTrend(
  storeId: string,
  days?: number
): Promise<RecoveryDailyCount[]> {
  return apiGet<RecoveryDailyCount[]>('/api/v1/recovery/analytics/trend', {
    params: { store_id: storeId, ...(days ? { days: String(days) } : {}) },
  });
}

export async function getRecoverySequences(
  storeId: string,
  options?: { page?: number; pageSize?: number }
): Promise<PaginatedResponse<RecoverySequence>> {
  return apiGet<PaginatedResponse<RecoverySequence>>(
    '/api/v1/recovery/sequences',
    {
      params: {
        store_id: storeId,
        ...(options?.page ? { page: String(options.page) } : {}),
        ...(options?.pageSize
          ? { page_size: String(options.pageSize) }
          : {}),
      },
    }
  );
}

export async function stopRecoverySequence(
  storeId: string,
  sequenceId: string
): Promise<{ status: string }> {
  return apiPost<{ status: string }>(
    `/api/v1/recovery/sequences/${sequenceId}/stop`,
    {
      params: { store_id: storeId },
    }
  );
}

export async function getAbandonedCheckouts(
  storeId: string,
  options?: { page?: number; pageSize?: number }
): Promise<PaginatedResponse<AbandonedCheckout>> {
  return apiGet<PaginatedResponse<AbandonedCheckout>>(
    '/api/v1/recovery/checkouts',
    {
      params: {
        store_id: storeId,
        ...(options?.page ? { page: String(options.page) } : {}),
        ...(options?.pageSize
          ? { page_size: String(options.pageSize) }
          : {}),
      },
    }
  );
}

export async function getRecoverySettings(
  storeId: string
): Promise<RecoverySettings> {
  return apiGet<RecoverySettings>('/api/v1/recovery/settings', {
    params: { store_id: storeId },
  });
}

export async function updateRecoverySettings(
  storeId: string,
  data: RecoverySettings
): Promise<RecoverySettings> {
  return apiPatch<RecoverySettings>('/api/v1/recovery/settings', {
    params: { store_id: storeId },
    body: data,
  });
}

// === Query Keys ===

export const recoveryKeys = {
  all: ['recovery'] as const,
  summary: (storeId: string, days?: number) =>
    [...recoveryKeys.all, 'summary', storeId, days] as const,
  trend: (storeId: string, days?: number) =>
    [...recoveryKeys.all, 'trend', storeId, days] as const,
  sequences: (storeId: string, page?: number) =>
    [...recoveryKeys.all, 'sequences', storeId, page] as const,
  checkouts: (storeId: string, page?: number) =>
    [...recoveryKeys.all, 'checkouts', storeId, page] as const,
  settings: (storeId: string) =>
    [...recoveryKeys.all, 'settings', storeId] as const,
};

File: apps/api/app/schemas/recovery.py
"""Pydantic schemas for M4 Cart Recovery."""

import re
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, field_validator

from app.schemas.common import BaseSchema

# --- Store recovery settings ---


class RecoverySettings(BaseSchema):
    """Store-level recovery configuration (stored in Store.settings['recovery'])."""

    enabled: bool = False
    min_cart_value: float = 0.0
    abandonment_threshold_minutes: int = 60
    sequence_timing_minutes: list[int] = Field(default_factory=lambda: [120, 1440, 2880, 4320])
    discount_enabled: bool = False
    discount_percent: int = 10
    max_emails_per_day: int = 50
    exclude_email_patterns: list[str] = Field(default_factory=list)

    @field_validator("exclude_email_patterns")
    @classmethod
    def validate_email_patterns(cls, patterns: list[str]) -> list[str]:
        for pattern in patterns:
            if len(pattern) > 200:
                raise ValueError(f"Pattern too long (max 200 chars): {pattern[:50]!r}...")
            try:
                re.compile(pattern)
            except re.error as exc:
                raise ValueError(f"Invalid regex pattern {pattern!r}: {exc}") from exc
        return patterns


# --- Abandoned checkout responses ---


class AbandonedCheckoutResponse(BaseSchema):
    """API response for an abandoned checkout."""

    id: UUID
    shopify_checkout_id: str
    customer_email: str | None
    customer_name: str | None
    total_price: float
    currency: str
    line_items: list[dict[str, Any]]
    checkout_url: str | None
    status: str
    abandonment_detected_at: datetime | None
    recovered_at: datetime | None
    created_at: datetime
    updated_at: datetime


# --- Recovery sequence responses ---


class RecoverySequenceResponse(BaseSchema):
    """API response for a recovery sequence."""

    id: UUID
    abandoned_checkout_id: UUID
    customer_email: str
    sequence_type: str
    status: str
    current_step_index: int
    steps_completed: list[dict[str, Any]]
    total_steps: int
    next_step_at: datetime | None
    started_at: datetime
    completed_at: datetime | None
    created_at: datetime


# --- Widget recovery check ---


class RecoveryItem(BaseSchema):
    """A single cart item for the widget recovery popup."""

    title: str
    price: str
    image_url: str | None = None
    quantity: int = 1


class RecoveryCheckResponse(BaseSchema):
    """Response for widget recovery popup check."""

    has_recovery: bool = False
    items: list[RecoveryItem] = Field(default_factory=list)
    checkout_url: str | None = None
    total_price: str | None = None
    sequence_id: UUID | None = None


# --- Recovery analytics ---


class RecoverySummary(BaseSchema):
    """Summary statistics for recovery analytics."""

    total_abandoned: int
    total_recovered: int
    recovery_rate: float
    recovered_revenue: float
    emails_sent: int
    active_sequences: int
    period_days: int


class RecoveryDailyCount(BaseSchema):
    """Daily count for recovery trend data."""

    date: str
    abandoned: int
    recovered: int
